contracts/two_prisoners_dilemma/lib.rs

```diff
@@ -0,0 +1,527 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+/// Smart Contract modelling classic Prisoner's Dilemma problem.
+/// To receive rewards, the contract needs to be pre-founded in the first place
+#[ink::contract]
+mod two_prisoners_dilemma {
+    use ink_lang::utils::initialize_contract;
+    use ink_prelude::string::{String, ToString};
+    use ink_storage::{
+        traits::{PackedLayout, SpreadAllocate, SpreadLayout},
+        Mapping,
+    };
+    use scale::{Decode, Encode};
+
+    // For now we harcode number of players to two, to model original dillema
+    const NUMBER_OF_PLAYERS: usize = 2;
+    pub const TOKEN_DECIMALS: u32 = 12;
+    pub const TOKEN: u128 = 10u128.pow(TOKEN_DECIMALS);
+
+    /// A choice of a player during the game
+    /// Initially all registered players are in `Waiting` state
+    #[derive(Debug, Encode, Decode, Clone, Copy, SpreadLayout, PackedLayout, PartialEq, Eq)]
+    #[cfg_attr(
+        feature = "std",
+        derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+    )]
+    enum Choice {
+        Waiting,
+        Cooperate,
+        Defect,
+    }
+
+    /// An event emitted when a reward was claimed and transfer succeeded
+    /// From acocunt is always the contract one
+    #[ink(event)]
+    pub struct Transfer {
+        #[ink(topic)]
+        to: AccountId,
+        value: Balance,
+    }
+
+    /// An event emitted when a player makes a legitimate choice
+    /// legitimate: not duplicate, and from registered players only
+    /// Obviously the exact choice is not revealed
```

Comment:
The whole problem with this game is that the choice is actually revealed. It's stored in the contract storage and everybody can go there and see it. Not sure how that could be circumvented, but in the current form the game does not make much sense.

It's still fine if we're ok with the game being pointless and the sole purpose of this contract is to be a tutorial, but in that case some work needs to be done wrt storage optimisation (currently you're storing much more data than needed)

Reply:
I think you attacked quite a hard problem here, where the task was simpler imo - ([rewrite popular ethereum contract(s) in ink!](https://cardinal-cryptography.atlassian.net/browse/A0-1190)).

IMHO writing:
- an escrow
- a blind auction
- a multisig wallet
- some defi "primitives" like a stablecoin
- one of the popular standards not yet implementen by OpenBrush (like ERC-777 or ERC-1155)

would be a) simpler, with lots of existing materials to draw inspiration from b) more "educational" for peeps looking to switch from Sol -> ink!

Reply:
To be honest it was not my intention to give out plain choices in the store. I realized this the same day afternoon after I published this PR. This is a serious flaw in the current versions when tokens are sent as rewards. Nevertheless, the code treated as an example is still a good use case of `ink!` features, such as storage, events, token transfer from contract, `ink!` unit tests, so I think it make sense to publish this example in either form:
1. Just change comments so that it is clearly said this is a non-production SC with transparent votes, only for tutorial purposes, or
2. Put some more effort and add hashing of the vote as per @fbielejec comment below, to get at least non-transparent votes. This version would be more or less similar to Blinded Auction (https://docs.soliditylang.org/en/v0.8.16/solidity-by-example.html#id2). 

My intention also was to come up with a new idea other than that already presented in https://github.com/paritytech/ink/tree/master/examples , which contains already name server(https://github.com/paritytech/ink/tree/master/examples/dns) , nevertheless but the list you provided looks very interesting. 

What do you think we still go with this PR with version 1. or 2., and I'll prepare another PR with a classic example like Escrow or MultiSig wallet?